exchange/exchange.go

```diff
@@ -208,27 +208,28 @@ func (e *exchange) buildBidResponse(ctx context.Context, liveAdapters []openrtb_
 	targData.addTargetsToCompletedAuction(auc)
 	bidResponse.SeatBid = seatBids
 
-	bidResponseExt := e.makeExtBidResponse(adapterBids, adapterExtra, bidRequest.Test, errList)
+	bidResponseExt := e.makeExtBidResponse(adapterBids, adapterExtra, bidRequest, errList)
 	ext, err := json.Marshal(bidResponseExt)
 	bidResponse.Ext = ext
 	return bidResponse, err
 }
 
 // Extract all the data from the SeatBids and build the ExtBidResponse
-func (e *exchange) makeExtBidResponse(adapterBids map[openrtb_ext.BidderName]*pbsOrtbSeatBid, adapterExtra map[openrtb_ext.BidderName]*seatResponseExtra, test int8, errList []error) *openrtb_ext.ExtBidResponse {
+func (e *exchange) makeExtBidResponse(adapterBids map[openrtb_ext.BidderName]*pbsOrtbSeatBid, adapterExtra map[openrtb_ext.BidderName]*seatResponseExtra, req *openrtb.BidRequest, errList []error) *openrtb_ext.ExtBidResponse {
 	bidResponseExt := &openrtb_ext.ExtBidResponse{
 		Errors:             make(map[openrtb_ext.BidderName][]string, len(adapterBids)),
 		ResponseTimeMillis: make(map[openrtb_ext.BidderName]int, len(adapterBids)),
 	}
-	if test == 1 {
+	if req.Test == 1 {
 		bidResponseExt.Debug = &openrtb_ext.ExtResponseDebug{
-			HttpCalls: make(map[openrtb_ext.BidderName][]*openrtb_ext.ExtHttpCall),
+			HttpCalls:       make(map[openrtb_ext.BidderName][]*openrtb_ext.ExtHttpCall),
+			ResolvedRequest: req,
```

Comment:
hmmm...... the only thing I'm a bit worried about, here, is that the same `BidRequest` instance was passed around as a pointer through a _lot_ of code before it got here. I think you've definitely done the most elegant thing, here... but I'm a little bit worried that it'll get mutated somewhere during the auction logic and end up returning a bad result here.

Although it's uglier, it'd probably be safer to `json.Marshal` the object as soon as you get it, and pass it through here as a `json.RawMessage`. You could also make a deep copy of the request struct, but... that's a lot of performance overhead, and you'd need to add a new project dependency (since deep clone isn't in Go's core libs), so... not sure it's worth all that trouble.

Reply:
I think marshaling it to JSON (if `req.Test == 1`) at the beginning of `HoldAuction` (or somewhere before `getAllBids`) should be sufficient to avoid risk of mutation by adapters, and then passing that down through `buildBidResponse` and `makeExtBidResponse` to here. It seems a bit awkward having the debug value extracted so far from where it's placed in the response, though I don't see a better way at the moment.